### PR TITLE
Support spring.config.import

### LINKED ISSUE #2266: Nacos: support 'spring.config.import'
Nacos

Spring Boot 2.4 introduced a new way to import configuration data via the `spring.config.import` property, and Spring Cloud 2020.0.x (consul) already support.

i see the Spring-Cloud-Alibaba `2021.1` is the first version that supports the 2020 version of spring cloud, but can not found about `spring.config.import`.

What time Spring-Cloud-Alibaba (nacos) will follow up and support it?

thanks

### PR DESCRIPTION
**Pull request by DanielLiu1123**
### Describe what this PR does / why we need it
Support for spring.config.Import to introduce NACOS configuration, no longer need Bootstrap container.

支持 spring.config.import 的方式引入 nacos 配置, 不再需要 bootstrap 容器

### Does this pull request fix one issue?
Fixed #2266  #2345

### How to use it
```yaml
<dependency>
    <groupId>com.alibaba.cloud</groupId>
    <artifactId>spring-cloud-starter-alibaba-nacos-config</artifactId>
    <!-- 排除 bootstrap, 未来版本 spring-cloud-alibaba 应该在 spring boot >= 2.4.0 时将该依赖设置为 optional -->
    <exclusions>
        <exclusion>
            <groupId>org.springframework.cloud</groupId>
            <artifactId>spring-cloud-starter-bootstrap</artifactId>
        </exclusion>
    </exclusions>
</dependency>
```
```yaml
# application.yml (不能是bootstrap.yml)
spring:
  cloud:
    nacos:
      config:
        group: DEFAULT_GROUP
        server-addr: localhost:8848
  config:
    import:
      - optional:nacos:test.yml  # 监听 DEFAULT_GROUP:test.yml
      - optional:nacos:test01.yml?group=group_01 # 覆盖默认 group, 监听 group_01:test01.yml
      - optional:nacos:test02.yml?group=group_02&refreshEnabled=false # 不开启动态刷新
      - nacos:test03.yml # 在拉取nacos配置异常时会快速失败, 会导致 spring 容器启动失败
```
等价于 (springboot 2.4.0 以下版本)
```yaml
# bootstrap.yml
spring:
  cloud:
    nacos:
      config:
        name: test.yml
        file-extension: yml
        group: DEFAULT_GROUP
        server-addr: localhost:8848
        extension-configs:
          - dataId: test01.yml
            group: group_01
          - dataId: test02.yml
            group: group_02
            refresh: false
```

### Precautions
1. spring boot >= 2.4.0 默认不会启动 bootstrap 容器, 不要将上述配置写在 bootstrap.yml(properties) 里
2. dataId 不可缺省, 不要使用相同 dataId

### CODE REVIEW

File: spring-cloud-alibaba-starters/spring-cloud-starter-alibaba-nacos-config/src/main/java/com/alibaba/cloud/nacos/configdata/NacosConfigDataLocationResolver.java [hunk 1]
```diff
@@ -0,0 +1,298 @@
...
+
+		return locations;
+	}
+
+	private void registerOrUpdateIndexes(String uris, NacosConfigProperties properties,
+			ConfigurableBootstrapContext bootstrapContext) {
+		String namespace = Objects.toString(properties.getNamespace(), DEFAULT_NAMESPACE);
+		Properties prop = new Properties();
+		prop.put(PropertyKeyConst.NAMESPACE, namespace);
+		prop.put(PropertyKeyConst.SERVER_ADDR, serverAddrFor(uris));
```
Comment (yuhuangbin) [inline spring-cloud-alibaba-starters/spring-cloud-starter-alibaba-nacos-config/src/main/java/com/alibaba/cloud/nacos/configdata/NacosConfigDataLocationResolver.java:164]: support more configuration items?

File: spring-cloud-alibaba-starters/spring-cloud-starter-alibaba-nacos-config/src/main/java/com/alibaba/cloud/nacos/configdata/NacosConfigDataLocationResolver.java [hunk 2]
```diff
@@ -0,0 +1,298 @@
...
+			ConfigServiceIndexes indexes = bootstrapContext
+					.get(ConfigServiceIndexes.class);
+			if (!indexes.getIndexes().containsKey(namespace)) {
+				try {
+					indexes.getIndexes().put(namespace,
+							ConfigFactory.createConfigService(prop));
+				}
+				catch (NacosException ignored) {
+				}
+			}
```
Comment (yuhuangbin) [inline spring-cloud-alibaba-starters/spring-cloud-starter-alibaba-nacos-config/src/main/java/com/alibaba/cloud/nacos/configdata/NacosConfigDataLocationResolver.java:175]: `putIfAbsent()` is better?

File: spring-cloud-alibaba-starters/spring-cloud-starter-alibaba-nacos-config/src/main/java/com/alibaba/cloud/nacos/configdata/NacosConfigDataLocationResolver.java (hunk 1 above)
Comment (DanielLiu1123) [reply spring-cloud-alibaba-starters/spring-cloud-starter-alibaba-nacos-config/src/main/java/com/alibaba/cloud/nacos/configdata/NacosConfigDataLocationResolver.java]: fixed

File: spring-cloud-alibaba-starters/spring-cloud-starter-alibaba-nacos-config/src/main/java/com/alibaba/cloud/nacos/configdata/NacosConfigDataLocationResolver.java (hunk 2 above)
Comment (DanielLiu1123) [reply spring-cloud-alibaba-starters/spring-cloud-starter-alibaba-nacos-config/src/main/java/com/alibaba/cloud/nacos/configdata/NacosConfigDataLocationResolver.java]: fixed

File: spring-cloud-alibaba-starters/spring-cloud-starter-alibaba-nacos-config/src/main/java/com/alibaba/cloud/nacos/configdata/NacosConfigDataLocationResolver.java [hunk 3]
```diff
@@ -0,0 +1,295 @@
...
+		ConfigService configService;
+		try {
+			configService = ConfigFactory.createConfigService(properties.assembleConfigServiceProperties());
+		} catch (NacosException e) {
+			return;
+		}
+		if (bootstrapContext.isRegistered(ConfigServiceIndexes.class)) {
+			ConfigServiceIndexes indexes = bootstrapContext
+					.get(ConfigServiceIndexes.class);
+			indexes.getIndexes().putIfAbsent(namespace, configService);
```
Comment (yuhuangbin) [inline spring-cloud-alibaba-starters/spring-cloud-starter-alibaba-nacos-config/src/main/java/com/alibaba/cloud/nacos/configdata/NacosConfigDataLocationResolver.java:213]: In fact, namespace is the only one, `ConfigServiceIndexes `is not very useful?

Comment (DanielLiu1123) [reply spring-cloud-alibaba-starters/spring-cloud-starter-alibaba-nacos-config/src/main/java/com/alibaba/cloud/nacos/configdata/NacosConfigDataLocationResolver.java]: yep, refactored

File: spring-cloud-alibaba-starters/spring-cloud-starter-alibaba-nacos-config/src/main/java/com/alibaba/cloud/nacos/configdata/NacosBootstrapper.java [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+import org.springframework.boot.BootstrapRegistry;
+import org.springframework.boot.BootstrapRegistry.InstanceSupplier;
+import org.springframework.boot.BootstrapRegistryInitializer;
+import org.springframework.boot.context.config.ConfigData;
+import org.springframework.boot.context.config.ConfigDataLoaderContext;
+import org.springframework.boot.context.properties.bind.Binder;
+import org.springframework.util.Assert;
+
+/**
+ * Spring boot >= 2.4.5 supports {@link BootstrapRegistryInitializer}.
```
Comment (zkzlx) [inline spring-cloud-alibaba-starters/spring-cloud-starter-alibaba-nacos-config/src/main/java/com/alibaba/cloud/nacos/configdata/NacosBootstrapper.java:33]: when the user boot version is relatively low, if they directly upgrade to your current version, will it be compatible?
e.g:boot version is 2.1.x or 2.2.x , the nacos-config  is latest version

Comment (DanielLiu1123) [reply spring-cloud-alibaba-starters/spring-cloud-starter-alibaba-nacos-config/src/main/java/com/alibaba/cloud/nacos/configdata/NacosBootstrapper.java:33]: This question is a bit confusing, if user uses boot version is 2.1.x or 2.2.x, he should not use SCA 2021 version.

Do you mean if boot version < 2.4.5 is compatible?

If it's this question, then the answer is compatible. This is an extension class, just do not have custom extensions when the version is less than 2.4.5.

File: spring-cloud-alibaba-starters/spring-cloud-starter-alibaba-nacos-config/src/main/java/com/alibaba/cloud/nacos/configdata/NacosConfigDataLocationResolver.java (hunk 3 above)
Comment (yuhuangbin) [reply spring-cloud-alibaba-starters/spring-cloud-starter-alibaba-nacos-config/src/main/java/com/alibaba/cloud/nacos/configdata/NacosConfigDataLocationResolver.java]: use `NacosConfigManager` create `ConfigService` ?

Comment (DanielLiu1123) [reply spring-cloud-alibaba-starters/spring-cloud-starter-alibaba-nacos-config/src/main/java/com/alibaba/cloud/nacos/configdata/NacosConfigDataLocationResolver.java]: better choice, fixed

File: spring-cloud-alibaba-starters/spring-cloud-starter-alibaba-nacos-config/src/main/java/com/alibaba/cloud/nacos/configdata/NacosBootstrapper.java (hunk 4 above)
Comment (zkzlx) [reply spring-cloud-alibaba-starters/spring-cloud-starter-alibaba-nacos-config/src/main/java/com/alibaba/cloud/nacos/configdata/NacosBootstrapper.java:33]: all current features are based on 2.3.x ;if the SCA version is upgraded to the latest , the feature cannot be used and cannot even be started if the boot version is lower than 2.4.5 ；
if user does not upgrade the boot version to 2.4.5 , his application may be not start.
i still suggest you be compatible this.

Comment (yuhuangbin) [reply spring-cloud-alibaba-starters/spring-cloud-starter-alibaba-nacos-config/src/main/java/com/alibaba/cloud/nacos/configdata/NacosBootstrapper.java:33]: The pr supported feature is submitted to Spring Cloud Alibaba 2021.x based on SpringBoot 2.4.x, so users won't exist in versions lower than SpringBoot 2.4.x?

Comment (DanielLiu1123) [reply spring-cloud-alibaba-starters/spring-cloud-starter-alibaba-nacos-config/src/main/java/com/alibaba/cloud/nacos/configdata/NacosBootstrapper.java:33]: In fact, I have been testing, user does not have problems when the version is below 2.4.5, and the application can start and work properly.

The current SCA use boot 2.4.2, currently remove this feature, when SCA use boot 2.4.5 and above, then add this feature ?  I think this is a solution.

Comment (DanielLiu1123) [reply spring-cloud-alibaba-starters/spring-cloud-starter-alibaba-nacos-config/src/main/java/com/alibaba/cloud/nacos/configdata/NacosBootstrapper.java]: Fixed

File: spring-cloud-alibaba-starters/spring-cloud-starter-alibaba-nacos-config/src/main/java/com/alibaba/cloud/nacos/configdata/NacosConfigDataLocationResolver.java [hunk 5]
```diff
@@ -0,0 +1,250 @@
...
+			path = "/" + path;
+		}
+		return properties.getServerAddr() + path;
+	}
+
+	private void registerConfigService(NacosConfigProperties properties,
+									   ConfigurableBootstrapContext bootstrapContext) {
+		if (!bootstrapContext.isRegistered(ConfigService.class)) {
+			Optional.ofNullable(new NacosConfigManager(properties).getConfigService())
+					.ifPresent(configService -> bootstrapContext.register(
```
Comment (yuhuangbin) [inline spring-cloud-alibaba-starters/spring-cloud-starter-alibaba-nacos-config/src/main/java/com/alibaba/cloud/nacos/configdata/NacosConfigDataLocationResolver.java]: not `new NacosConfigManager` to get `ConfigService`. 
I think you should register The `NacosConfigManager` with the `ConfigurableBootstrapContext ` and then use the `NacosConfigManager#getConfigService()`

File: spring-cloud-alibaba-starters/spring-cloud-starter-alibaba-nacos-config/src/main/java/com/alibaba/cloud/nacos/configdata/NacosConfigDataLocationResolver.java [hunk 6]
```diff
@@ -0,0 +1,250 @@
...
+	}
+
+	private URI getUri(String uris) {
+		if (!uris.startsWith("http://") && !uris.startsWith("https://")) {
+			uris = "http://" + uris;
+		}
+		URI uri;
+		try {
+			uri = new URI(uris);
+		} catch (URISyntaxException e) {
```
Comment (yuhuangbin) [inline spring-cloud-alibaba-starters/spring-cloud-starter-alibaba-nacos-config/src/main/java/com/alibaba/cloud/nacos/configdata/NacosConfigDataLocationResolver.java]: `getUri()` Repeatedly invoked, It is recommended to build once and then pass through parameters.

File: spring-cloud-alibaba-starters/spring-cloud-starter-alibaba-nacos-config/src/main/java/com/alibaba/cloud/nacos/configdata/NacosConfigDataLocationResolver.java [hunk 7]
```diff
@@ -0,0 +1,250 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.alibaba.cloud.nacos.configdata;
+
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.util.*;
+
```
Comment (yuhuangbin) [inline spring-cloud-alibaba-starters/spring-cloud-starter-alibaba-nacos-config/src/main/java/com/alibaba/cloud/nacos/configdata/NacosConfigDataLocationResolver.java:26]: not `import *`

File: spring-cloud-alibaba-starters/spring-cloud-starter-alibaba-nacos-config/src/main/java/com/alibaba/cloud/nacos/configdata/NacosConfigDataLocationResolver.java (hunk 5 above)
Comment (DanielLiu1123) [reply spring-cloud-alibaba-starters/spring-cloud-starter-alibaba-nacos-config/src/main/java/com/alibaba/cloud/nacos/configdata/NacosConfigDataLocationResolver.java]: Fixed

File: spring-cloud-alibaba-starters/spring-cloud-starter-alibaba-nacos-config/src/main/java/com/alibaba/cloud/nacos/configdata/NacosConfigDataLocationResolver.java (hunk 6 above)
Comment (DanielLiu1123) [reply spring-cloud-alibaba-starters/spring-cloud-starter-alibaba-nacos-config/src/main/java/com/alibaba/cloud/nacos/configdata/NacosConfigDataLocationResolver.java]: Fixed

File: spring-cloud-alibaba-starters/spring-cloud-starter-alibaba-nacos-config/src/main/java/com/alibaba/cloud/nacos/configdata/NacosConfigDataLocationResolver.java (hunk 7 above)
Comment (DanielLiu1123) [reply spring-cloud-alibaba-starters/spring-cloud-starter-alibaba-nacos-config/src/main/java/com/alibaba/cloud/nacos/configdata/NacosConfigDataLocationResolver.java:26]: Fixed

**Review by yuhuangbin — APPROVED** [review]
LGTM

**Comment by zhadan142857** [comment]
```
14:47:27.012 [main] ERROR org.springframework.boot.SpringApplication - Application run failed
java.lang.IllegalStateException: Unable to load config data from 'optional:dev:test.yaml?group=DEFAULT_GROUP'
	at org.springframework.boot.context.config.StandardConfigDataLocationResolver.getReferences(StandardConfigDataLocationResolver.java:141)
	at org.springframework.boot.context.config.StandardConfigDataLocationResolver.getReferences(StandardConfigDataLocationResolver.java:126)
	at org.springframework.boot.context.config.StandardConfigDataLocationResolver.resolve(StandardConfigDataLocationResolver.java:119)
	at org.springframework.boot.context.config.ConfigDataLocationResolvers.lambda$resolve$1(ConfigDataLocationResolvers.java:115)
	at org.springframework.boot.context.config.ConfigDataLocationResolvers.resolve(ConfigDataLocationResolvers.java:126)
	at org.springframework.boot.context.config.ConfigDataLocationResolvers.resolve(ConfigDataLocationResolvers.java:115)
	at org.springframework.boot.context.config.ConfigDataLocationResolvers.resolve(ConfigDataLocationResolvers.java:107)
	at org.springframework.boot.context.config.ConfigDataImporter.resolve(ConfigDataImporter.java:105)
	at org.springframework.boot.context.config.ConfigDataImporter.resolve(ConfigDataImporter.java:97)
	at org.springframework.boot.context.config.ConfigDataImporter.resolveAndLoad(ConfigDataImporter.java:85)
	at org.springframework.boot.context.config.ConfigDataEnvironmentContributors.withProcessedImports(ConfigDataEnvironmentContributors.java:116)
	at org.springframework.boot.context.config.ConfigDataEnvironment.processInitial(ConfigDataEnvironment.java:240)
	at org.springframework.boot.context.config.ConfigDataEnvironment.processAndApply(ConfigDataEnvironment.java:227)
	at org.springframework.boot.context.config.ConfigDataEnvironmentPostProcessor.postProcessEnvironment(ConfigDataEnvironmentPostProcessor.java:102)
	at org.springframework.boot.context.config.ConfigDataEnvironmentPostProcessor.postProcessEnvironment(ConfigDataEnvironmentPostProcessor.java:94)
	at org.springframework.boot.env.EnvironmentPostProcessorApplicationListener.onApplicationEnvironmentPreparedEvent(EnvironmentPostProcessorApplicationListener.java:102)
	at org.springframework.boot.env.EnvironmentPostProcessorApplicationListener.onApplicationEvent(EnvironmentPostProcessorApplicationListener.java:87)
	at org.springframework.context.event.SimpleApplicationEventMulticaster.doInvokeListener(SimpleApplicationEventMulticaster.java:176)
	at org.springframework.context.event.SimpleApplicationEventMulticaster.invokeListener(SimpleApplicationEventMulticaster.java:169)
	at org.springframework.context.event.SimpleApplicationEventMulticaster.multicastEvent(SimpleApplicationEventMulticaster.java:143)
	at org.springframework.context.event.SimpleApplicationEventMulticaster.multicastEvent(SimpleApplicationEventMulticaster.java:131)
	at org.springframework.boot.context.event.EventPublishingRunListener.environmentPrepared(EventPublishingRunListener.java:85)
	at org.springframework.boot.SpringApplicationRunListeners.lambda$environmentPrepared$2(SpringApplicationRunListeners.java:66)
	
```

**Comment by sochina-heart** [comment]
nacos如果开启认证鉴权，并且nacos密码使用jasypt进行加密后，使用spring.config.import这种引入配置文件，会报错，会先加载nacos配置，再进行解密，这个有什么好的解决方法吗

**Pull request merged**